File: kaasb-step5/backend/app/schemas/contract.py
"""
Kaasb Platform - Contract & Milestone Schemas
"""

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


# === Embedded Info ===


class ContractUserInfo(BaseModel):
    """Minimal user info in contract responses."""
    id: uuid.UUID
    username: str
    first_name: str
    last_name: str
    display_name: Optional[str] = None
    avatar_url: Optional[str] = None

    model_config = {"from_attributes": True}


class ContractJobInfo(BaseModel):
    """Minimal job info in contract responses."""
    id: uuid.UUID
    title: str
    category: str
    job_type: str

    model_config = {"from_attributes": True}


# === Milestone Schemas ===


class MilestoneCreate(BaseModel):
    """Create a milestone on a contract."""
    title: str = Field(min_length=3, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    amount: float = Field(ge=1, le=100000)
    due_date: Optional[datetime] = None
    order: int = Field(ge=0, le=100, default=0)


class MilestoneUpdate(BaseModel):
    """Update a pending milestone."""
    title: Optional[str] = Field(None, min_length=3, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    amount: Optional[float] = Field(None, ge=1, le=100000)
    due_date: Optional[datetime] = None
    order: Optional[int] = Field(None, ge=0, le=100)


class MilestoneSubmit(BaseModel):
    """Freelancer submits a milestone for review."""
    submission_note: Optional[str] = Field(None, max_length=2000)


class MilestoneReview(BaseModel):
    """Client reviews a submitted milestone."""
    action: str = Field(pattern=r"^(approve|request_revision)$")
    feedback: Optional[str] = Field(None, max_length=2000)


class MilestoneDetail(BaseModel):
    """Full milestone response."""
    id: uuid.UUID
    title: str
    description: Optional[str] = None
    amount: float
    order: int
    status: str
    due_date: Optional[datetime] = None
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None
    paid_at: Optional[datetime] = None
    submission_note: Optional[str] = None
    feedback: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# === Contract Schemas ===


class ContractCreate(BaseModel):
    """Client creates milestones when setting up a contract."""
    milestones: list[MilestoneCreate] = Field(min_length=1, max_length=20)


class ContractDetail(BaseModel):
    """Full contract response with milestones."""
    id: uuid.UUID
    title: str
    description: Optional[str] = None
    total_amount: float
    amount_paid: float
    status: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    deadline: Optional[datetime] = None
    created_at: datetime

    client: ContractUserInfo
    freelancer: ContractUserInfo
    job: ContractJobInfo
    milestones: list[MilestoneDetail] = []

    model_config = {"from_attributes": True}


class ContractSummary(BaseModel):
    """Contract card in listings."""
    id: uuid.UUID
    title: str
    total_amount: float
    amount_paid: float
    status: str
    started_at: datetime

    client: ContractUserInfo
    freelancer: ContractUserInfo
    job: ContractJobInfo

    milestone_count: int = 0
    completed_milestones: int = 0

    model_config = {"from_attributes": True}


class ContractListResponse(BaseModel):
    """Paginated contract list."""
    contracts: list[ContractSummary]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/contract.ts
// === Contract Types ===

export interface ContractUserInfo {
  id: string;
  username: string;
  first_name: string;
  last_name: string;
  display_name?: string;
  avatar_url?: string;
}

export interface ContractJobInfo {
  id: string;
  title: string;
  category: string;
  job_type: string;
}

export interface MilestoneDetail {
  id: string;
  title: string;
  description?: string;
  amount: number;
  order: number;
  status: string;
  due_date?: string;
  submitted_at?: string;
  approved_at?: string;
  paid_at?: string;
  submission_note?: string;
  feedback?: string;
  created_at: string;
}

export interface ContractDetail {
  id: string;
  title: string;
  description?: string;
  total_amount: number;
  amount_paid: number;
  status: string;
  started_at: string;
  completed_at?: string;
  deadline?: string;
  created_at: string;
  client: ContractUserInfo;
  freelancer: ContractUserInfo;
  job: ContractJobInfo;
  milestones: MilestoneDetail[];
}

export interface ContractSummary {
  id: string;
  title: string;
  total_amount: number;
  amount_paid: number;
  status: string;
  started_at: string;
  client: ContractUserInfo;
  freelancer: ContractUserInfo;
  job: ContractJobInfo;
  milestone_count: number;
  completed_milestones: number;
}

export interface ContractListResponse {
  contracts: ContractSummary[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// === Status display helpers ===

export const CONTRACT_STATUS_LABELS: Record<string, string> = {
  active: "نشط",
  completed: "مكتمل",
  cancelled: "ملغى",
  disputed: "متنازع عليه",
  paused: "موقوف",
};

export const CONTRACT_STATUS_COLORS: Record<string, string> = {
  active: "bg-blue-50 text-blue-700 border-blue-200",
  completed: "bg-green-50 text-green-700 border-green-200",
  cancelled: "bg-gray-100 text-gray-500 border-gray-200",
  disputed: "bg-red-50 text-red-700 border-red-200",
  paused: "bg-yellow-50 text-yellow-700 border-yellow-200",
};

export const MILESTONE_STATUS_LABELS: Record<string, string> = {
  pending: "معلق",
  in_progress: "جارٍ",
  submitted: "مُقدَّم",
  revision_requested: "طلب مراجعة",
  approved: "موافق عليه",
  paid: "مدفوع",
};

export const MILESTONE_STATUS_COLORS: Record<string, string> = {
  pending: "bg-gray-100 text-gray-600",
  in_progress: "bg-blue-50 text-blue-700",
  submitted: "bg-purple-50 text-purple-700",
  revision_requested: "bg-orange-50 text-orange-700",
  approved: "bg-green-50 text-green-700",
  paid: "bg-green-100 text-green-800",
};
